### PR TITLE
Use a dynamic RESTMapper for admission plugins

### PR DESCRIPTION
**Pull request by enj**
**What this PR does / why we need it**:

This change updates the REST mapper used by all admission plugins to be backed by cached discovery information.  This cache is updated every ten seconds via a post start hook and will not attempt to update on calls to `RESTMapping`.  It solely relies on the hook to keep the cache in sync with discovery.

This prevents issues with the `OwnerReferencesPermissionEnforcement` admission plugin when it is used with custom resources that set `blockOwnerDeletion`.

**Which issue(s) this PR fixes**:
`Fixes #...`

**Special notes for your reviewer**:

There are probably other ways the post start hook could be wired.

**Release note**:

```release-note
NONE
```

Signed-off-by: Monis Khan <mkhan@redhat.com>

@kubernetes/sig-api-machinery-misc

### CODE REVIEW

File: test/integration/examples/apiserver_test.go [hunk 1]
```diff
@@ -141,7 +141,7 @@ func TestAggregatedAPIServer(t *testing.T) {
 	}()
 
 	// just use json because everyone speaks it
-	err = wait.PollImmediate(100*time.Millisecond, 10*time.Second, func() (done bool, err error) {
+	err = wait.PollImmediate(time.Second, time.Minute, func() (done bool, err error) {
```
Comment (deads2k) [inline test/integration/examples/apiserver_test.go]: Why the change?

File: cmd/kube-apiserver/app/server.go [hunk 2]
```diff
@@ -582,15 +589,22 @@ func BuildAdmissionPluginInitializers(
...
+	restMapper := discovery.NewDeferredDiscoveryRESTMapper(discoveryClient, meta.InterfacesForUnstructured)
 
 	quotaConfiguration := quotainstall.NewQuotaConfigurationForAdmission()
 
 	kubePluginInitializer := kubeapiserveradmission.NewPluginInitializer(client, sharedInformers, cloudConfig, restMapper, quotaConfiguration)
 	webhookPluginInitializer := webhookinit.NewPluginInitializer(webhookAuthWrapper, serviceResolver)
 
-	return []admission.PluginInitializer{webhookPluginInitializer, kubePluginInitializer}, nil
+	return []admission.PluginInitializer{webhookPluginInitializer, kubePluginInitializer},
+		func(context genericapiserver.PostStartHookContext) error {
```
Comment (deads2k) [inline cmd/kube-apiserver/app/server.go]: assign to variable near the restmapper so it's easier to read.

Comment (enj) [reply cmd/kube-apiserver/app/server.go]: ok

File: test/integration/examples/apiserver_test.go (hunk 1 above)
Comment (enj) [reply test/integration/examples/apiserver_test.go]: It was flaking on my machine when it was under load (so I synced it with the values in the ETCD test).

**Comment by enj** [comment]
Not sure how to fix `test-update-storage-objects.sh` (`client.Discovery()` returns `nil` for it):

```
1: running [Created by server.(*GenericAPIServer).RunPostStartHooks @ hooks.go:151]
    runtime   runtime.go:58     HandleCrash(0, 0, 0)
              panic.go:502      panic(0x32d6540, 0x8590280)
    cached    memcache.go:142   (*memCacheClient).Invalidate(0xc420483340)
    discovery restmapper.go:219 (*DeferredDiscoveryRESTMapper).Reset(0xc420302300)
    app       server.go:603     BuildAdmissionPluginInitializers.func1(#3, #1, 0, 0)
    server    hooks.go:184      runPostStartHook.func1(#2, #4, #3, #1, 0xc4218d9f60)
    server    hooks.go:185      runPostStartHook(0x3ad0823, 0x2a, #2, #4, #3, #1)
```

```
// KUBE_API_VERSIONS is used in test-update-storage-objects.sh, disabling a number of API
// groups. This leads to a nil client above and undefined behaviour further down.
//
// TODO: get rid of KUBE_API_VERSIONS or define sane behaviour if set
glog.Errorf("Failed to create clientset with KUBE_API_VERSIONS=%q: %v. KUBE_API_VERSIONS is only for testing. Things will break.",
	kubeAPIVersions, err)
```

File: cmd/kube-apiserver/app/server.go [hunk 3]
```diff
@@ -597,7 +597,7 @@ func BuildAdmissionPluginInitializers(
 	}
 
 	// We have a functional client so we can use that to build our discovery backed REST mapper
-	if client != nil {
+	if client != nil && client.Discovery() != nil {
```
Comment (deads2k) [inline cmd/kube-apiserver/app/server.go]: squash

Comment (enj) [reply cmd/kube-apiserver/app/server.go]: Done
